File: app/src/main/java/com/kakeibo/AmountTextWatcher.java
package com.kakeibo;

import android.text.Editable;
import android.text.TextWatcher;
import android.util.Log;
import android.widget.EditText;

import java.util.HashSet;
import java.util.Set;

public class AmountTextWatcher implements TextWatcher {
    private EditText mEdtAmount;
    private int mFractionDigits;

    public AmountTextWatcher(EditText edtAmount, int fractionDigits) {
        mEdtAmount = edtAmount;
        mFractionDigits = fractionDigits;
    }

    @Override
    public void beforeTextChanged(CharSequence charSequence, int i, int i1, int i2) {}

    @Override
    public void onTextChanged(CharSequence s, int start, int before, int count) { }

    @Override
    public void afterTextChanged(Editable editable) {
        String str = editable.toString();
        int length = str.length();

        if (mFractionDigits == 0) {
            if (length > 1 && str.charAt(length - 1) == '.') {
                mEdtAmount.setText(str.substring(0, length - 1));
                mEdtAmount.setSelection(mEdtAmount.getText().length());
            }
            return;
        }

        if (length == 1) {
            if (str.charAt(0) == '.') {
                mEdtAmount.setText("");
            }
        } else if (length > 1) {
            if (str.charAt(0) == '0' && str.charAt(1) != '.') {
                mEdtAmount.setText(str.substring(0, length - 1));
                mEdtAmount.setSelection(mEdtAmount.getText().length());
                return;
            }

            if (str.charAt(length - 1) == '.' && secondTime(str)) {
                mEdtAmount.setText(str.substring(0, length - 1));
                mEdtAmount.setSelection(mEdtAmount.getText().length());
                return;
            }

            if (str.contains(".") &&
                    str.substring(str.indexOf('.')).length() > mFractionDigits + 1) {
                mEdtAmount.setText(str.substring(0, length - 1));
                mEdtAmount.setSelection(mEdtAmount.getText().length());
            }
        }
    }

    private boolean secondTime(String str) {
        Set<Character> set = new HashSet<>();

        for (int i = 0; i < str.length(); ++i) {
            if (set.contains(str.charAt(i)) && str.charAt(i)=='.') {
                return true;
            }
            set.add(str.charAt(i));
        }
        return false;
    }
}

File: app/src/main/java/com/kakeibo/ui/view/AmountTextWatcher.kt
//package com.kakeibo.ui.view
//
//import android.text.Editable
//import android.text.TextWatcher
//import android.widget.EditText
//import com.kakeibo.R
//import com.kakeibo.SubApp
//
//class AmountTextWatcher(private val edtAmount: EditText) : TextWatcher {
//
//    private val _fractionDigits = SubApp.getFractionDigits(R.string.pref_key_fraction_digits)
//
//    override fun beforeTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {}
//    override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {}
//    override fun afterTextChanged(editable: Editable) {
//
//        val str = editable.toString()
//        val length = str.length
//
//        if (_fractionDigits == 0) {
//            if (length >= 1 && (str[length - 1] == '.' || str[length - 1] == ',')) {
//                edtAmount.setText(str.substring(0, length - 1))
//                edtAmount.setSelection(edtAmount.text.length)
//            }
//            if (length >= 1 && str[0] == '0') {
//                edtAmount.setText("")
//            }
//            return
//        }
//
//        if (length > 1) {
//            // consecutive 0s from the beginning is not allowed
//            if (str[0] == '0' && str[1] == '0') {
//                edtAmount.setText(str.substring(0, length - 1))
//                edtAmount.setSelection(edtAmount.text.length)
//            }
//            // containing two periods, commas, a period after a comma, a comma after a period are not allowed
//            if (str[length - 1] == '.' && (str.substring(0, length - 1).contains('.')
//                        || str.substring(0, length - 1).contains(','))) {
//                edtAmount.setText(str.substring(0, length - 1))
//                edtAmount.setSelection(edtAmount.text.length)
//                return
//            }
//            if (str[length - 1] == ',' && (str.substring(0, length - 1).contains('.')
//                        || str.substring(0, length - 1).contains(','))) {
//                edtAmount.setText(str.substring(0, length - 1))
//                edtAmount.setSelection(edtAmount.text.length)
//                return
//            }
//            // handling digits part
//            if (str.contains(".") &&
//                str.substring(str.indexOf('.')).length > _fractionDigits + 1) {
//                edtAmount.setText(str.substring(0, length - 1))
//                edtAmount.setSelection(edtAmount.text.length)
//            }
//            // handling digits part
//            if (str.contains(",") &&
//                str.substring(str.indexOf(',')).length > _fractionDigits + 1) {
//                edtAmount.setText(str.substring(0, length - 1))
//                edtAmount.setSelection(edtAmount.text.length)
//            }
//        }
//    }
//}